### ISSUE TITLE
Support CUDA 10.1 Update 2

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#2569]
Currently CuPy does not work with CUDA 10.1 Update 2. It should be resolved by the final release of v7.

* Following tests are failing (
    * cupy_tests.fft_tests.test_fft.TestRfftn_param_11.test_irfftn
    * tests.cupy_tests.linalg_tests.test_product.TestMatrixPower.test_matrix_power_inv2

* #2542 broke compatibility with CUDA 10.1 Update 2. Some undocumented APIs (e.g.,  `cusparseCbsrilu02_bufferSizeExt`) are dropped in CUDA 10.1 Update 2. As they are still declared in `cusparse.h` (this is a bug of cuSPARSE), build succeeds but `import cupy` fails. (PR #2573)
```
CUDA 10.1.105-1: ok
CUDA 10.1.168-1: ok
CUDA 10.1.243-1: fail (cuSPARSE 10.3.0.243)
```

* CUDA 10.1 Update 2 now uses semantic versioning, it is better to print version of every library (cusolver, cusparse, ...) currently linked to CuPy in `cupy.show_config`. (PR #2578)
* CUDA 10.1 Update 2 changed FP16 headers (PR #2592)

**Comment by jakirkham** [comment]
So this an issue with the current release or just `master`?

**Comment by kmaehashi** [comment]
v7.0.0rc1 release and current `master` branches are affected.

**Comment by anaruse** [comment]
> #2542 broke compatibility with CUDA 10.1 Update 2. Some undocumented APIs

Sorry, this was my mistake. I was using CUDA 10.1 Update 1 at that time and did not notice this. I checked that some undocumented APIs like `cusparseCbsrilu02_bufferSizeExt` are actually removed from libcusparse.so.10.3.0.243, though those are still in libcusparse_static.a.

**Comment by kmaehashi** [comment]
Test results:

> cupy_tests.fft_tests.test_fft.TestRfftn_param_11.test_irfftn
https://jenkins.preferred.jp/job/chainer/job/chainer-test_pr/70/label=mn1-p100,test=cupy-combination/testReport/junit/tests.cupy_tests.fft_tests.test_fft/TestRfftn_param_11/test_irfftn/

> tests.cupy_tests.linalg_tests.test_product.TestMatrixPower.test_matrix_power_inv2
https://jenkins.preferred.jp/job/chainer/job/chainer-test_pr/70/label=mn1-p100,test=cupy-combination/testReport/junit/tests.cupy_tests.linalg_tests.test_product/TestMatrixPower/test_matrix_power_inv2/

**Comment by emcastillo** [comment]
The API calls that #2573 removed were not present in the stable release 😃

**Comment by kmaehashi** [comment]
Test failures are currently under investigation by @anaruse.

**Comment by anaruse** [comment]
Sorry for being very late to respond.

> cupy_tests.fft_tests.test_fft.TestRfftn_param_11.test_irfftn

The issue above does not reproduce on my environment (CUDA 10.1 U2, GPU V100). Could you try it on V100 as well?

> tests.cupy_tests.linalg_tests.test_product.TestMatrixPower.test_matrix_power_inv2

This issue occurs on my environment. Seems this is partly because results of `testing.shaped_arange((3, 3), xp, dtype) ** 2`, that is used as an input array `a` in `xp.linalg.matrix_power(a, -2)`, are a little bit strange in cupy with CUDA 10.1 U2 at least.

    >>> import cupy
    >>> from cupy import testing
    >>> testing.shaped_arange((3, 3), cupy, cupy.float32) ** 2
    array([[ 1.      ,  4.      ,  9.      ],
       [16.      , 24.999998, 36.      ],
       [48.999996, 64.      , 81.      ]], dtype=float32)

And there is no discrepancy when `a` is computed in following way

    >>> import cupy
    >>> from cupy import testing
    >>> a = testing.shaped_arange((3, 3), cupy, cupy.float32)
    >>> a * a
    array([[ 1.,  4.,  9.],
           [16., 25., 36.],
           [49., 64., 81.]], dtype=float32)

Any thoughts on this?

**Comment by anaruse** [comment]
> tests.cupy_tests.linalg_tests.test_product.TestMatrixPower.test_matrix_power_inv2

Seems the test above raises an AssertionError with not only CUDA 10.1 U2 but also 10.1 U1 (no error with CUDA 10.1).  When the test is changed as follows, no error is raised even with CUDA 10.1 U1 or 10.1 U2.

    -        a = testing.shaped_arange((3, 3), xp, dtype) ** 2
    +        a = testing.shaped_arange((3, 3), xp, dtype)
    +        a = a * a

It looks there are at least two small accuracy issue: one in `cupy.power`, another in `cupy.linalg.matrix_power`.

**Comment by anaruse** [comment]
The accuracy issue of `cupy.power`, when float32 is used, can be fixed by the following change. Though it might cause some performance degradation...

    --- a/cupy/core/_routines_math.pyx
    +++ b/cupy/core/_routines_math.pyx
    @@ -516,7 +516,7 @@ _power = create_ufunc(
         ('??->b', 'bb->b', 'BB->B', 'hh->h', 'HH->H', 'ii->i', 'II->I', 'll->l',
          'LL->L', 'qq->q', 'QQ->Q',
          ('ee->e', 'out0 = powf(in0, in1)'),
    -     ('ff->f', 'out0 = powf(in0, in1)'),
    +     ('ff->f', 'out0 = (float)pow((double)in0, (double)in1)'),
          ('dd->d', 'out0 = pow(in0, in1)'),
          ('FF->F', 'out0 = pow(in0, in1)'),
          ('DD->D', 'out0 = pow(in0, in1)')),

**Comment by anaruse** [comment]
FYI

i | powf(i,2) on CPU | powf(i,2) on GPU
-- | -- | --
0 |  0.000000 |  0.000000
1 |  1.000000 |  1.000000
2 |  4.000000 |  4.000000
3 |  9.000000 |  9.000000
4 | 16.000000 | 16.000000
5 | 25.000000 | 24.999998
6 | 36.000000 | 36.000000
7 | 49.000000 | 48.999996
8 | 64.000000 | 64.000000
9 | 81.000000 | 81.000000

**Comment by kmaehashi** [comment]
>> cupy_tests.fft_tests.test_fft.TestRfftn_param_11.test_irfftn

> The issue above does not reproduce on my environment (CUDA 10.1 U2, GPU V100). Could you try it on V100 as well?

OK, I'll try. FYI, I reproduced this error on P100 with CUDA 10.1. Also reproducible with GTX 1060 using CUDA 10.2 (cuFFT 10102).

**Comment by grlee77** [comment]
On current master with CUDA 10.1 update 2, I am also seeing a two test failures in `cupyx.scipy.fft` that I assume are for the same reason as the `test_irfftn` one mentioned above. Specifically they are:
```
Results (36.41s):
    1126 passed
       2 failed
         - tests/cupyx_tests/scipy_tests/fft_tests/test_fft.py:363 TestRfftn_param_11_{axes=(-1, -2, -3), norm=None, s=None, shape=(2, 3, 4)}.test_irfftn
         - tests/cupyx_tests/scipy_tests/fft_tests/test_fft.py:374 TestRfftn_param_11_{axes=(-1, -2, -3), norm=None, s=None, shape=(2, 3, 4)}.test_irfftn_overwrite
```

**Comment by anaruse** [comment]
> cupy_tests.fft_tests.test_fft.TestRfftn_param_11.test_irfftn

The issue was reproduced here as well. It seems the issue happens when
* Pascal generation GPU is used,
* Inverse (complex-to-real) FFT is performed,
* Size of array in last transform axis is two, and
* CUDA version is 10.1 update 2 or 10.2.

**Comment by anaruse** [comment]
https://github.com/cupy/cupy/blob/f06f509167957240d17924965238cb0901b700ab/tests/cupy_tests/fft_tests/test_fft.py#L631-L633

Now, the test above uses a randomly created array as an input to `irfftn`. Would it be OK to use an output of `rfftn` instead as the input to `irfftn`?

**Comment by anaruse** [comment]
Re-opened. Sorry, I closed this issue by mistake ..

**Comment by anaruse** [comment]
> The issue was reproduced here as well. It seems the issue happens when
> * Pascal generation GPU is used,
> * Inverse (complex-to-real) FFT is performed,
> * Size of array in last transform axis is two, and
> * CUDA version is 10.1 update 2 or 10.2.

This issue will be fixed in a future CUDA release.

**Comment by leofang** [comment]
> > The issue was reproduced here as well. It seems the issue happens when
> > 
> > * Pascal generation GPU is used,
> > * Inverse (complex-to-real) FFT is performed,
> > * Size of array in last transform axis is two, and
> > * CUDA version is 10.1 update 2 or 10.2.
> 
> This issue will be fixed in a future CUDA release.

@anaruse Any chance you can check if this bug is fixed in CUDA 11.0 or not?
